design/cstor/v1/pool_expansion_by_expanding_disk.md

```diff
@@ -0,0 +1,145 @@
+---
+
+# Pool Expansion When Underlying Disk Expanded
+
+## Table of Contents
+
+- [Pool Expansion When Underlying Disk Expanded](#pool-expansion-when-underlying-disk-expanded)
+	- [Table of Contents](#table-of-contents)
+	- [Summary](#summary)
+	- [Motivation](#motivation)
+		- [Goals](#goals)
+		- [Non-Goals](#non-goals)
+	- [Proposal](#proposal)
+		- [User Stories](#user-stories)
+		- [Current Implementation](#current-implementation)
+		- [Proposed Implementations](#proposed-implementations)
+		- [Steps to perform user stories](#steps-to-perform-user-stories)
+			- [Pool Expansion Steps:](#pool-expansion-steps)
+		- [Low Level Design](#low-level-design)
+			- [Work Flow](#work-flow)
+	- [Drawbacks](#drawbacks)
+	- [Alternatives](#alternatives)
+	- [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+- This proposal includes design details of the pool expansion when underlying disks were expanded.
+
+## Motivation
+
+There were cases where OpenEBS users expand the pool by expanding underlying disks(which were used by pool) when space on the pool got exhausted.
+
+### Goals
+
+- Expand cStor pool when the underlying disk got expanded.
+
+### Non-Goals
+
+NA
+
+## Proposal
+
+### User Stories
+
+As a OpenEBS user I should be able to expand the pool when underlying disk got expanded.
+
+### Current Implementation
+
+When user creates the CSPC, CSPC controller which is watching for CSPC will get create request and creats cstorpoolinstances and cstorpooldeployments. CstorPoolInstance is responsible for holding the blockdevice information and other pool configurations while cStor pool deployment contains following containers cstor-pool-mgmt, cstor-pool container and m-exporter as name conveyes cstor-pool-mgmt is responsible for managing the pool. cstor-pool container creates pool(ZFS file system) on top of the blockdevices. If any changes need to make on pool configuration like adding blockdevices, adding raid groups and replacing blockdevice cstor-pool-mgmt container will look into CstorPoolInstance for pool information and verifies are there any pending actions to be performed then cstor-pool-mgmt will prforms corresponding operation on the pool. Below info is hight level info to know how the operation is detected
+
+- Pool Expansion When blockdevice/raid group added:	
+	cstor-pool-mgmt will iterate over all the blockdevices present in CstorPoolInstance and then verifies whether corresponding blockdevice is in use in the pool. If that blockdevice/raid group doesn’t exist then trigger is for pool expansion by adding blockdevice/raid group.
+
+- BlockDevice replacement:
+	cstor-pool-mgmt will iterate over all the blockdevices present in CstorPoolInstance and the verifies is that blockdevice is for replacement(We are maintaining annotation called openebs.io/predecessor: <replaced_blockdevice_name> on the claim of that blockdevice) if that block device is for replacement then the it will perform replacement operation on the pool.
+
```

Comment:
cstor-pool manager will iterate over all the `BlockDevices` present in `CstorPoolInstance` and verifies the requested `BlockDevice` replacement request (we are maintaining annotation called `openebs.io/predecessor: <replaced_blockdevice_name>` on the claim of that blockdevice).  if  requested `BlockDevice` is valid for replacement then manger will perform `BlockDevice` replacement operation on the pool.